cloud-controller-manager/do/firewall_controller_test.go

```diff
@@ -0,0 +1,420 @@
+						return []godo.Firewall{*test.expectedListResult}, test.expectedListResponse, test.expectedListError
+					},
+					addRulesFunc: func(ctx context.Context, fID string, rr *godo.FirewallRulesRequest) (*godo.Response, error) {
+						return test.expectedAddRulesResponse, test.expectedAddRulesError
+					},
+					updateFunc: func(ctx context.Context, fID string, fr *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+						return test.expectedUpdateResult, test.expectedUpdateResponse, test.expectedUpdateError
+					},
+					createFunc: func(ctx context.Context, fr *godo.FirewallRequest) (*godo.Firewall, *godo.Response, error) {
+						return test.expectedCreateResult, test.expectedCreateResponse, test.expectedCreateError
+					},
+					getFunc: func(ctx context.Context, fID string) (*godo.Firewall, *godo.Response, error) {
+						return test.expectedGetResult, test.expectedGetResponse, test.expectedGetError
+					},
+				},
+			)
+			fwManagerOp = newFakeFirewallManagerOp(gclient, test.fwCache)
+			fc := NewFirewallController(ctx, kclient, gclient, inf.Core().V1().Services(), fwManagerOp, []string{})
+
+			err := fc.fwManager.Set(ctx, test.inboundRules)
+			if want, got := test.expectedError, err; !reflect.DeepEqual(want, got) {
+				t.Errorf("incorrect firewall config\nwant: %#v\n got: %#v", want, got)
+			}
+		})
+	}
+}
+
+func TestFirewallController_createInboundRules(t *testing.T) {
+	fakeWorkerFirewallName := "myFirewallWorkerName"
+	inboundRule := godo.InboundRule{
+		Protocol:  "tcp",
+		PortRange: "31220",
+	}
+	nodePortService := &v1.Service{
+		Spec: v1.ServiceSpec{
+			Type: v1.ServiceTypeNodePort,
+			Ports: []v1.ServicePort{
+				{
+					Protocol: "tcp",
+					Port:     31220,
+					NodePort: 31220,
+				},
+			},
+		},
+	}
+	fakeServiceList := []*v1.Service{nodePortService}
+
+	testcases := []struct {
+		name        string
+		firewall    *godo.Firewall
+		fwCache     firewallCache
+		serviceList []*v1.Service
+	}{
+		{
+			name:        "successfully updates port range",
+			firewall:    nil,
+			fwCache:     newFakeFirewallCache(fakeWorkerFirewallName, inboundRule),
+			serviceList: fakeServiceList,
+		},
+		{
+			name:        "return node port inbound rules when we find a service of type nodeport",
+			firewall:    newFakeFirewall(fakeWorkerFirewallName, inboundRule),
+			fwCache:     newFakeFirewallCache(fakeWorkerFirewallName, inboundRule),
+			serviceList: fakeServiceList,
+		},
+	}
+
+	for _, test := range testcases {
+		t.Run(test.name, func(t *testing.T) {
+			// setup
+			gclient := newFakeGodoClient(
+				&fakeFirewallService{},
+			)
+			inboundRules := []godo.InboundRule{inboundRule}
+			fwManagerOp = newFakeFirewallManagerOp(gclient, test.fwCache)
+			fc := NewFirewallController(ctx, kclient, gclient, inf.Core().V1().Services(), fwManagerOp, []string{})
+
+			rules := fc.createInboundRules(test.serviceList)
+			cmp.Equal(rules[0].PortRange, inboundRules[0].PortRange)
```

Comment:
This needs to be wrapped in / called by an assertion -- a return value of `false` from `cmp.Equal` won't let the test fail per se.